// react-poppop/lib/portal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var hasCreatePortal = _reactDom2.default.createPortal !== undefined;
var createPortal = hasCreatePortal ? _reactDom2.default.createPortal : _reactDom2.default.unstable_renderSubtreeIntoContainer;

var Portal = function (_Component) {
  _inherits(Portal, _Component);

  function Portal(props) {
    _classCallCheck(this, Portal);

    var _this = _possibleConstructorReturn(this, (Portal.__proto__ || Object.getPrototypeOf(Portal)).call(this, props));

    _this.el = document.createElement('div');
    document.body.appendChild(_this.el);
    _this.originBodyOverflow = document.body.style.overflow;
    _this.isHidden = _this.originBodyOverflow === 'hidden';

    if (!_this.isHidden) {
      document.body.style.overflow = 'hidden';
    }
    return _this;
  }

  _createClass(Portal, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.renderUnstableSubtree();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.renderUnstableSubtree();
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (!nextProps.open) {
        this.resetBodyOverflow();
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.removePortal();
      this.resetBodyOverflow();
    }
  }, {
    key: 'resetBodyOverflow',
    value: function resetBodyOverflow() {
      if (!this.isHidden) {
        document.body.style.overflow = this.originBodyOverflow;
      }
    }
  }, {
    key: 'removePortal',
    value: function removePortal() {
      if (hasCreatePortal) {
        document.body.removeChild(this.el);
      } else {
        _reactDom2.default.unmountComponentAtNode(this.el);
      }
    }
  }, {
    key: 'renderUnstableSubtree',
    value: function renderUnstableSubtree() {
      if (!hasCreatePortal) {
        _reactDom2.default.unstable_renderSubtreeIntoContainer(this, this.props.children, this.el);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      if (hasCreatePortal) {
        return _reactDom2.default.createPortal(this.props.children, this.el);
      }
      return null;
    }
  }]);

  return Portal;
}(_react.Component);

exports.default = Portal;
module.exports = exports['default'];
